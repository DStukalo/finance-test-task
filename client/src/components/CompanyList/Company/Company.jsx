/* eslint-disable react-hooks/exhaustive-deps */
import { PropTypes } from 'prop-types';
import { useEffect } from 'react';
import { useLastPrices } from '../../../store/lastPricesStore';
import useRunningLineFacade from '../../../hooks/useRunningStoreFacade';
import useRunningLineStatusFacade from '../../../hooks/useRunningLineStatusFacade';
import styles from './Company.module.scss';

export function Company({
	name, price, percent, change, fullname,
}) {
	const lastPrices = useLastPrices();
	const runLine = useRunningLineFacade();
	const runLineStatus = useRunningLineStatusFacade();
	const difference = parseFloat(price) > parseFloat(lastPrices[`Last${name}`]);

	function onClickHandler() {
		if (runLine[fullname]) runLine[`set${fullname}False`]();
		else runLine[`set${fullname}True`]();
	}

	useEffect(() => {
		if (difference) runLineStatus[`set${fullname}Up`]();
		else runLineStatus[`set${fullname}Down`]();
	}, [difference]);

	useEffect(
		() => () => lastPrices[`setLast${name}`](price),
		[price],
	);

	return (
		<div className={styles.company}>
			<div className={styles.company__brand}>
				{fullname !== '' && <div className={styles[fullname]} />}
				<div className={styles.company__name}>
					{fullname}
				</div>
			</div>
			{price !== '' ? (
				<p className={styles.company__price}>
					{price}
					$
				</p>
			) : null}
			{difference ? (
				<>
					<div className={styles.company__up}>
						<p className={styles.company__percent}>
							{percent}
							%
						</p>
						<div className={styles.company__up_logo} />
					</div>
					<div className={styles.company__change_up}>
						+
						{change}
						$
					</div>
				</>
			) : (
				<>
					<div className={styles.company__down}>
						<p>
							{percent}
							%
						</p>
						<div className={styles.company__down_logo} />
					</div>
					<div className={styles.company__change_down}>
						-
						{change}
						$
					</div>
				</>
			)}
			{!runLine[fullname] ? (
				<button type="button" className={styles.company__btn} onClick={onClickHandler}>+</button>
			) : (
				<button type="button" className={styles.company__btn} onClick={onClickHandler}>-</button>
			)}

		</div>
	);
}

Company.propTypes = {
	name: PropTypes.string,
	price: PropTypes.string,
	percent: PropTypes.string,
	change: PropTypes.string,
	fullname: PropTypes.string,
};

Company.defaultProps = {
	name: '',
	price: '',
	percent: '',
	change: '',
	fullname: '',
};
