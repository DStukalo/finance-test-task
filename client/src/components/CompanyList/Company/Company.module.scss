.company {
	padding: 10px;
	width: 500px;
	height: 60px;
	display: flex;
	justify-content: space-between;
	align-items: center;
	border-top: 1px solid var(--color-black-a1);
	font-size: var(--font-size-a4);

	@media (max-width: 520.99px) {
		width: max(95%, 300px);
		font-size: var(--font-size-a5);
	}

	&__name,
	&__price {
		width: 55px;
	}

	&__brand {
		display: flex;
		align-items: center;
		gap: 5px;
	}

	&__up {
		padding: 5px;
		width: 80px;
		height: 30px;
		border-radius: 5px;
		display: flex;
		align-items: center;
		background-color: var(--color-green-a1);
	}

	&__up_logo {
		width: 32px;
		height: 24px;
		background-image: url("../../../../public/arrow_up.png");
		background-position: center;
		background-size: contain;
		background-repeat: no-repeat;
	}

	&__down {
		padding: 5px;
		width: 80px;
		height: 30px;
		border-radius: 5px;
		display: flex;
		justify-content: center;
		align-items: center;
		background-color: var(--color-violet-a1);
	}

	&__down_logo {
		width: 32px;
		height: 24px;
		background-image: url("../../../../public/arrow_down.png");
		background-position: center;
		background-size: contain;
		background-repeat: no-repeat;
	}

	&__change_up {
		padding: 5px;
		width: 75px;
		border-radius: 5px;
		text-align: center;
		background-color: var(--color-green-a1);

		@media (max-width: 520.99px) {
			display: none;
		}
	}

	&__change_down {
		padding: 5px;
		width: 75px;
		border-radius: 5px;
		text-align: center;
		background-color: var(--color-violet-a1);

		@media (max-width: 520.99px) {
			display: none;
		}
	}

	&__btn {
		padding: 5px;
		width: 30px;
		background-color: inherit;
		border-radius: 50%;
		font-size: var(--font-size-a2);
		color: var(--color-orange-a2);
		border: 2px solid var(--color-orange-a2);
	}
}

.Facebook {
	width: 24px;
	height: 24px;
	background-image: url(../../../../public/assets/facebook_png.png);
	background-size: cover;
}

.Apple {
	width: 24px;
	height: 24px;
	background-image: url(../../../../public/assets/apple_icon.svg);
	background-size: cover;
}

.Alphabet {
	width: 24px;
	height: 24px;
	background-image: url(../../../../public/assets/google_icon.svg);
}

.Microsoft {
	width: 24px;
	height: 24px;
	background-image: url(../../../../public/assets/microsoft_icon.svg);
}

.Amazon {
	width: 24px;
	height: 24px;
	background-image: url(../../../../public/assets/amazon_icon.svg);
}

.Tesla {
	width: 24px;
	height: 24px;
	background-image: url(../../../../public/assets/tesla_icon.svg);
}
