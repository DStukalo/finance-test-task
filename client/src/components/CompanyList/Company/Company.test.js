import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Company } from './Company';

describe('Company component', () => {
	it('Company render with fullname Apple', () => {
		render(<Company fullname="Apple" name="AAPL" price="137.33" percent="0.35" change="171.57" />, { wrapper: BrowserRouter });

		expect(screen.getByText('Apple')).toHaveClass('company__name');

		expect(screen.queryByText('Alphabet')).not.toBeInTheDocument();
		expect(screen.queryByText('Microsoft')).not.toBeInTheDocument();
		expect(screen.queryByText('Amazon')).not.toBeInTheDocument();
		expect(screen.queryByText('Facebook')).not.toBeInTheDocument();
		expect(screen.queryByText('Tesla')).not.toBeInTheDocument();
	});

	it('render 6 Company component', () => {
		render(<Company fullname="Apple" name="AAPL" price="137.33" percent="0.35" change="171.57" />, { wrapper: BrowserRouter });
		render(<Company fullname="Alphabet" name="AAPL" price="137.33" percent="0.35" change="171.57" />, { wrapper: BrowserRouter });
		render(<Company fullname="Microsoft" name="AAPL" price="137.33" percent="0.35" change="171.57" />, { wrapper: BrowserRouter });
		render(<Company fullname="Amazon" name="AAPL" price="137.33" percent="0.35" change="171.57" />, { wrapper: BrowserRouter });
		render(<Company fullname="Facebook" name="AAPL" price="137.33" percent="0.35" change="171.57" />, { wrapper: BrowserRouter });
		render(<Company fullname="Tesla" name="AAPL" price="137.33" percent="0.35" change="171.57" />, { wrapper: BrowserRouter });

		expect(screen.getByText('Apple')).toHaveClass('company__name');
		expect(screen.getByText('Alphabet')).toHaveClass('company__name');
		expect(screen.getByText('Microsoft')).toHaveClass('company__name');
		expect(screen.getByText('Amazon')).toHaveClass('company__name');
		expect(screen.getByText('Facebook')).toHaveClass('company__name');
		expect(screen.getByText('Tesla')).toHaveClass('company__name');
	});

	it('Company snapshot', () => {
		const company = render(<Company fullname="Apple" name="AAPL" price="137.33" percent="0.35" change="171.57" />, { wrapper: BrowserRouter });
		expect(company).toMatchSnapshot();
	});
});
