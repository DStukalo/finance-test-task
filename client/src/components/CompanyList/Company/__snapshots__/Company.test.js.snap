// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company component Company snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="company"
      >
        <div
          class="company__brand"
        >
          <div
            class="Apple"
          />
          <div
            class="company__name"
          >
            Apple
          </div>
        </div>
        <p
          class="company__price"
        >
          137.33
          $
        </p>
        <div
          class="company__down"
        >
          <p>
            0.35
            %
          </p>
          <div
            class="company__down_logo"
          />
        </div>
        <div
          class="company__change_down"
        >
          -
          171.57
          $
        </div>
        <button
          class="company__btn"
          type="button"
        >
          +
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="company"
    >
      <div
        class="company__brand"
      >
        <div
          class="Apple"
        />
        <div
          class="company__name"
        >
          Apple
        </div>
      </div>
      <p
        class="company__price"
      >
        137.33
        $
      </p>
      <div
        class="company__down"
      >
        <p>
          0.35
          %
        </p>
        <div
          class="company__down_logo"
        />
      </div>
      <div
        class="company__change_down"
      >
        -
        171.57
        $
      </div>
      <button
        class="company__btn"
        type="button"
      >
        +
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
